File: src/my_project/medium_problems/from51to100/jump_game_dp.py
from typing import List 

class Solution:
    
    def canJump(self, nums: List[int]) -> bool:
        
        n = len(nums)
        
        if n==1:
            return True
        
        dp = [0]*(n)
        
        prev_best = n-1
        
        dp[n-1] = 1
        
        for i in range(n-2, -1, -1):
            
            step = nums[i]
            
            if (i+step)>=prev_best:
                
                dp[i] = 1
                prev_best = i
        
        
        if dp[0]:
            return True
        
        return False

File: my_project_typescript/medium_problems_typescript/from1to50/jumpGameDP.ts
function canJump(nums: number[]): boolean {

    let n: number = nums.length

    if (n === 1 ){
        return true
    }

    let dp: Array<number> = Array(n).fill(0)

    let prevBest: number = n-1 
    let step:number 
    dp[n-1] = 1 

    for (let i = n-2; i>-1; i--){
        step = nums[i]

        if ( i+step >= prevBest){
            dp[i] = 1 
            prevBest = i
        }
    }

    console.log(dp)

    if (dp[0] === 1){
        return true
    }
    return false
};

console.log(canJump([1,0,1,0]))